### PR TITLE
Improve resilience of remote stream restart

### LINKED ISSUE #14884: Improve resilience of remote stream restart
**Describe the bug**

The brokers currently naively request that all members restart their streams whenever a new member is added. This causes some noise since obviously the other brokers aren't handling such messages. 

Further, we don't retry failures, so we could still have streams that are not known on the brokers.

**Expected behavior**

- We only request stream restart from gateways.
- We retry (with back off) until the request is successful.

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR improves the resilience of the remote stream restart call. It adds error handling to the `RemoteStreamTransport#restartStreams` API. This API triggers newly added members (from the membership protocol) to send the broker their existing streams. It's required to ensure a consistent state on both gateways/brokers in the case where a broker would be restarted too fast, such that the gateways would not detect it ever left the membership group, and thus, would not register their existing streams on the new broker.

To avoid temporary network errors, the request is retried (with randomized exponential backoff) forever until one of the following cases is met:

- The remote member does not subscribe to the restart stream topic. This can happen when sending the request to, say, another broker. It can also happen with gateways if the gateway is still starting up or if it's shutting down. In all cases, it's fine to not retry and suppress the error.
- The remote member has since been removed from the membership protocol. If this happens, then we know either the remote member is dead, or it's temporarily unavailable. Either way, no point retrying until it's added back to the protocol, at which point `restartStreams` will be called again.
- The remote member threw an uncaught exception while handling the request. If the request nnot be processed, there's no point retrying.

Any other error - time outs, connection errors, etc. - will be retried infinitely, up to once per 5 seconds.

A few other minor things:

- `ConcurrencyControl` is now also an `Executor`
- I cloned the `ExponentialBackoff` utility from the client. We don't actually have a shared utility module between the client and the gateway/brokers, so I wasn't too sure where to put it, and didn't want to create a new module right now. Plus, I think eventually it would make sense to build retry/backoff/rate limit/etc. into our transport directly.
- On a similar note, I figured better copy that little utility than go for a heavy duty dependency like Resilience4J. We should look at that in the future, but for now it's that one class is fine I think.
- The tests use actual `AtomixCluster` only so they can get a real `ClusterCommunicationService`. Since we're testing the actual error behavior, I didn't feel comfortable mocking it with no easy way to later get integration tests for that.

## Related issues

closes #14884 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: transport/src/main/java/io/camunda/zeebe/transport/stream/impl/RemoteStreamTransport.java
Comment (npepinpe) [inline transport/src/main/java/io/camunda/zeebe/transport/stream/impl/RemoteStreamTransport.java:1]: :thought_balloon: At this point, there's enough logic that we might want to split out the "client" part out to its own class.

File: transport/src/main/java/io/camunda/zeebe/transport/stream/impl/RemoteStreamTransport.java [hunk 1]
```diff
@@ -69,45 +94,109 @@ public void removeAll(final MemberId member) {
...
+            membership protocol; can be safely ignored.""",
+            receiver,
+            e);
+        completed.complete(null);
+      }
+        // this error means the remote member is not handling requests of this type; either it's not
+        // a
+        // gateway, or it's still starting up or shutting down. in either case, restarting streams
+        // makes
+        // no sense
```
Comment (rodrigo-lourenco-lopes) [inline transport/src/main/java/io/camunda/zeebe/transport/stream/impl/RemoteStreamTransport.java:169]: very small format nitpick

```suggestion
        // this error means the remote member is not handling requests of this type; either it's not
        // a gateway, or it's still starting up or shutting down. in either case, restarting streams
        // makes no sense
```

**Review by rodrigo-lourenco-lopes — APPROVED** [review]
Looks good to me! 🙌

File: transport/src/main/java/io/camunda/zeebe/transport/stream/impl/RemoteStreamTransport.java (hunk 1 above)
Comment (npepinpe) [reply transport/src/main/java/io/camunda/zeebe/transport/stream/impl/RemoteStreamTransport.java:169]: Oops, that's when I pre-emptively put a breakline, then auto format :smile:

**Comment by npepinpe** [comment]
bors merge

**Pull request merged**